### DISCUSSION TITLE
How to register code snippets for a DSL?

### DISCUSSION THREAD

**Discussion by rhumbertgz** [eclipse-langium/langium#1118] [Q&A] [answered]
Hi. 

I'm using the MonacoEditorReactComp and I can't find a way to register my completion provider. Using the UserConfig I can specify the Monarch syntax highlighting, but there is no option to set my code snippets. Any pointers how to register them?

Best regards,
Humberto

**Comment by msujew** [comment]
Hey @rhumbertgz,

we don't have a builtin API for code snippets yet, but you can easily add them yourselves. See the following (inaccurate) example:

```ts
class CustomCompletionProvider extends DefaultCompletionProvider {
    override getCompletion(args): Promise<CompletionList> {
        const list = await super.getCompletion(args);
        list.items.push({
             label: 'snippet',
             kind: CompletionItemKind.Snippet,
             insertText: '...'
        });
        return list;
    }
}
```

And register that completion provider for your language.

**Reply by rhumbertgz** [reply]
Hi @msujew,

Thanks for your quick reaction and the example 👍🏼

**Reply by rhumbertgz** [reply]
Hi @msujew,

I managed to register the custom completion provider, but the code snippets don't works 100%. 

![Jul-12-2023 15-55-44](https://github.com/langium/langium/assets/10393245/578ebb7d-f838-4adf-a3b1-58523b64e7f5)

The DefaultCompletationProvider uses a CompletionItem from `vscode-languageserver` doesn't have the properties `range` and `insertTextRules`.  I was expecting to have a result similar to 👇🏼

![Jul-12-2023 15-53-29](https://github.com/langium/langium/assets/10393245/775453e9-6071-4428-8c1b-3c97bed43ae6)

Here is the implementation of my completion provider

//hello-world-completion.ts
```
import { DefaultCompletionProvider, LangiumDocument } from 'langium';
import {
  CompletionList,
  CompletionItemKind,
  CompletionParams,
  CompletionItem
} from 'vscode-languageserver';

type Suggestions = Promise<CompletionList | undefined>;


export class CustomCompletionProvider extends DefaultCompletionProvider {
  override async getCompletion(doc: LangiumDocument, params: CompletionParams): Suggestions {
    const list = await super.getCompletion(doc, params);
    console.log('getCompletion: ');
    if (list !== undefined) {
      const snippets: CompletionItem[]  = [
        {
          label: 'person',
          kind: CompletionItemKind.Snippet,
          insertText: 'person ${1:name}',
          documentation: 'Define a new person',
        },
        {
          label: 'hello',
          kind: CompletionItemKind.Snippet,
          insertText: 'hello ${1:person}!',
          documentation: 'Define a new greeting',
        }
      ]; 

      list.items.push(...snippets);
      console.log('list: ', list);
    } else {
      console.log('list undifined')
    }

    return list;
  }
}

```


Here is how I register the custom completion provider (not sure if it is the right way)
// hello-world-module.ts

```
...

export function createHelloWorldServices(context: DefaultSharedModuleContext): {
    shared: LangiumSharedServices,
    HelloWorld: HelloWorldServices
} {
   ...

    // Register Custom Completation Provider
  HelloWorld.lsp.CompletionProvider = new CustomCompletionProvider(HelloWorld);
  
    return { shared, HelloWorld };
}
```

Any suggestions?

I created a sandbox [here](https://codesandbox.io/p/github/rhumbertgz/hello-world/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cljzpm8rm000g3t6madexrjpe%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B66.8079280054135%252C33.1920719945865%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljzpm8rm000b3t6mbmznt5le%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%252C%2522activeTabId%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Flanguage%252Fhello-world-module.ts%2522%252C%2522id%2522%253A%2522cljzpo6rf019x3t6mmxysghgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cljzpm8rm000d3t6m13y4fqe9%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%252C%2522activeTabId%2522%253A%2522cljztibu2000b3d62ay18mhhj%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzr7uvp018r3t6m7scmzwk3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzrtmbq016c3t6mgr6hvuxs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522id%2522%253A%2522cljztibu2000b3d62ay18mhhj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522PROJECT_SETUP%2522%257D%255D%257D%252C%2522cljzpm8rm000f3t6mhwsxw2v0%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%252C%2522activeTabId%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljzpm8rm000e3t6mtftxitnj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljzr70qw0028fmil8q980e5a%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522id%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A19.153196839080465%257D).

**Reply by msujew** [reply, marked as answer]
I believe you're missing the `insertTextFormat: InsertTextFormat.Snippet` property in your returned items. I forgot that you need to set that as well.

**Reply by rhumbertgz** [reply]
Yes, that was missing. Everything works now. Thanks 👍🏼

**Reply by cdietrich** [reply]
kind?: CompletionItemKind is used for the icon only.

**Reply by rhumbertgz** [reply]
I updated the sandbox example, maybe it can help someone in the future  ;) 

[Sandbox demo](https://codesandbox.io/p/github/rhumbertgz/hello-world/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cljzpm8rm000g3t6madexrjpe%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B66.8079280054135%252C33.1920719945865%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljzpm8rm000b3t6mbmznt5le%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%252C%2522activeTabId%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Flanguage%252Fhello-world-module.ts%2522%252C%2522id%2522%253A%2522cljzpo6rf019x3t6mmxysghgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cljzpm8rm000d3t6m13y4fqe9%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%252C%2522activeTabId%2522%253A%2522cljztibu2000b3d62ay18mhhj%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzr7uvp018r3t6m7scmzwk3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzrtmbq016c3t6mgr6hvuxs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522id%2522%253A%2522cljztibu2000b3d62ay18mhhj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522PROJECT_SETUP%2522%257D%255D%257D%252C%2522cljzpm8rm000f3t6mhwsxw2v0%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%252C%2522activeTabId%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljzpm8rm000e3t6mtftxitnj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljzr70qw0028fmil8q980e5a%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522id%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A19.153196839080465%257D)
[Github repo](https://github.com/rhumbertgz/hello-world)
